src/libstd/io/mod.rs

```diff
@@ -0,0 +1,928 @@
+// leave `buf` in a state such that it contains invalid UTF-8 in its bounds.
+//
+// To this end, we use an RAII guard (to protect against panics) which updates
+// the length of the string when it is dropped. This guard initially truncates
+// the string to the prior length and only afer we've validated that the
+// new contents are valid UTF-8 do we allow it to set a longer length.
+//
+// The unsafety in this function is twofold:
+//
+// 1. We're looking at the raw bytes of `buf`, so we take on the burden of UTF-8
+//    checks.
+// 2. We're passing a raw buffer to the function `f`, and it is expected that
+//    the function only *appends* bytes to the buffer. We'll get undefined
+//    behavior if existing bytes are overwritten to have non-UTF-8 data.
+fn append_to_string<F>(buf: &mut String, f: F) -> Result<()>
+    where F: FnOnce(&mut Vec<u8>) -> Result<()>
+{
+    struct Guard<'a> { s: &'a mut Vec<u8>, len: usize }
+    #[unsafe_destructor]
+    impl<'a> Drop for Guard<'a> {
+        fn drop(&mut self) {
+            unsafe { self.s.set_len(self.len); }
+        }
+    }
+
+    unsafe {
+        let mut g = Guard { len: buf.len(), s: buf.as_mut_vec() };
+        let ret = f(g.s);
+        if str::from_utf8(&g.s[g.len..]).is_err() {
+            ret.and_then(|()| {
+                Err(Error::new(ErrorKind::InvalidInput,
+                               "stream did not contain valid UTF-8", None))
+            })
+        } else {
+            g.len = g.s.len();
+            ret
+        }
+    }
+}
+
+fn read_to_end<R: Read + ?Sized>(r: &mut R, buf: &mut Vec<u8>) -> Result<()> {
+    loop {
+        if buf.capacity() == buf.len() {
+            buf.reserve(DEFAULT_BUF_SIZE);
+        }
+        match with_end_to_cap(buf, |b| r.read(b)) {
+            Ok(0) => return Ok(()),
+            Ok(_) => {}
+            Err(ref e) if e.kind() == ErrorKind::Interrupted => {}
+            Err(e) => return Err(e),
+        }
+    }
+}
+
+/// A trait for objects which are byte-oriented sources.
+///
+/// Readers are defined by one method, `read`. Each call to `read` will attempt
+/// to pull bytes from this source into a provided buffer.
+///
+/// Readers are intended to be composable with one another. Many objects
+/// throughout the I/O and related libraries take and provide types which
+/// implement the `Read` trait.
+pub trait Read {
+    /// Pull some bytes from this source into the specified buffer, returning
+    /// how many bytes were read.
+    ///
+    /// This function does not provide any guarantees about whether it blocks
+    /// waiting for data, but if an object needs to block for a read but cannot
+    /// it will typically signal this via an `Err` return value.
+    ///
+    /// If the return value of this method is `Ok(n)`, then it must be
+    /// guaranteed that `0 <= n <= buf.len()`. If `n` is `0` then it indicates
+    /// that this reader as reached "end of file" and will likely no longer be
```

Comment:
There are well known examples where 0 does not mean that further reads will also return 0. Maybe this should not talk about what's "likely" or "unlikely".